wolfcrypt/src/sha.c

```diff
@@ -719,6 +719,14 @@ int wc_ShaFinal(wc_Sha* sha, byte* hash)
         return BAD_FUNC_ARG;
     }
 
+#ifndef WC_NO_HARDEN
```

Comment:
Please move all this down to where the `0x80` is set. Use `BUFFER_E` as the failure. Inline with the error check on the update. Please fix other algos to use this too.

```
% git diff
diff --git a/wolfcrypt/src/sha.c b/wolfcrypt/src/sha.c
index bcfd1005d..3037d7cab 100644
--- a/wolfcrypt/src/sha.c
+++ b/wolfcrypt/src/sha.c
@@ -719,8 +719,6 @@ int wc_ShaFinal(wc_Sha* sha, byte* hash)
         return BAD_FUNC_ARG;
     }

-    local = (byte*)sha->buffer;
-
 #ifdef WOLF_CRYPTO_CB
     if (sha->devId != INVALID_DEVID) {
         ret = wc_CryptoCb_ShaHash(sha, NULL, 0, hash);
@@ -737,6 +735,11 @@ int wc_ShaFinal(wc_Sha* sha, byte* hash)
     }
 #endif /* WOLFSSL_ASYNC_CRYPT */

+    if (sha->buffLen > WC_SHA_BLOCK_SIZE - 1) {
+        return BUFFER_E;
+    }
+
+    local = (byte*)sha->buffer;
     local[sha->buffLen++] = 0x80;  /* add 1 */

     /* pad with zeros */
```